modules/auxiliary/spoof/nbns/nbns_response.rb

```diff
@@ -44,108 +47,144 @@ def initialize
     ])
 
     register_advanced_options([
-      OptBool.new('Debug', [ false, "Determines whether incoming packet parsing is displayed", false])
+      OptBool.new('DEBUG', [ false, "Determines whether incoming packet parsing is displayed", false])
     ])
 
     deregister_options('RHOST', 'PCAPFILE', 'SNAPLEN', 'FILTER')
+    self.thread = nil
+    self.sock = nil
   end
 
-  def run
-    check_pcaprub_loaded() # Check first since otherwise this is all for naught
-    # MacOS X workaround
-    ::Socket.do_not_reverse_lookup = true
+  def dispatch_request(packet, rhost, src_port)
+    rhost = ::IPAddr.new(rhost)
+     # `recvfrom` (on Linux at least) will give us an ipv6/ipv4 mapped
+    # addr like "::ffff:192.168.0.1" when the interface we're listening
+    # on has an IPv6 address. Convert it to just the v4 addr
+    if rhost.ipv4_mapped?
+      rhost = rhost.native
+    end
 
-    @sock = ::UDPSocket.new()
-    @sock.setsockopt(::Socket::SOL_SOCKET, ::Socket::SO_REUSEADDR, 1)
-    @sock.bind('', 137) # couldn't specify srv host because it missed broadcasts
+    # Convert to string
+    rhost = rhost.to_s
 
-    @run = true
+    spoof = ::IPAddr.new(datastore['SPOOFIP'])
 
-    print_status("NBNS Spoofer started. Listening for NBNS requests...")
+    return if packet.length == 0
 
-    begin
+    nbnsq_transid      = packet[0..1]
+    nbnsq_flags        = packet[2..3]
+    nbnsq_questions    = packet[4..5]
+    nbnsq_answerrr     = packet[6..7]
+    nbnsq_authorityrr  = packet[8..9]
+    nbnsq_additionalrr = packet[10..11]
+    nbnsq_name         = packet[12..45]
+    decoded = ""
+    nbnsq_name.slice(1..-2).each_byte do |c|
+      decoded << "#{(c - 65).to_s(16)}"
+    end
+    nbnsq_decodedname = "#{[decoded].pack('H*')}".strip()
+    nbnsq_type         = packet[46..47]
+    nbnsq_class        = packet[48..49]
+
+    return unless nbnsq_decodedname =~ /#{datastore['REGEX']}/i
```

Comment:
@jhart-r7: agreed, in cases where we're not redefining the regex mid-way through, this is a considerable time saver. I ran across this when writing our MITM proxies (IIRC a bunch of that code lives on in closed PRs in this repo), and it makes a pretty significant difference when you're running full packet streams through MSF and doing something as taxing as failing most regex checks already - if there are lots of expressions and they're being compiled on the fly it can be unpleasant under MRI (2.1.5 seems better than 1.9.3, but RBX 2.5.2 actually handles those better in the long run since you can compile 'em all async and the JIT optimizes a bunch of the code path).

In cases where we're attempting to create intelligent MITM methods which redefine the match and sub on the fly, we can't avoid the recompilation.

Regex are expensive objects which fail more often than not in network environments, and someone smarter than me once said that "the worst thing a regular expression can do is fail to find something." As far as i know Ruby VMs do not natively support pattern matching like Scala does. However, there is a [pattern matching lib] (https://github.com/k-tsj/pattern-match) i'm going to take a look at in the coming days/weeks since it dovetails with commercial endeavors. Given the fact that the thing decomposes objects and we do scary things to instantiated objects with :extend calls all the time, i'd like to ask @hmoore-r7 to eyeball the lib as well to check for glaring problems.